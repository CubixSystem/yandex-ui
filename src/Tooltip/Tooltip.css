.Tooltip {
    z-index: 1500;

    font-family: var(--tooltip-font-family);
}

.Tooltip.Popup2::before {
    content: none;
}
