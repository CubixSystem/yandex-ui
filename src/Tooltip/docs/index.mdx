---
id: Tooltip
path: /components/tooltip

title: Tooltip
description: Компонент, использующийся для создания всплывающих подсказок.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Tooltip
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Tooltip.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1

tabs:
 - Examples
 - Usage
 - Props
 - Accessibility
---
