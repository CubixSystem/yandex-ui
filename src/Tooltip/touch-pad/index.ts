export * from '../Tooltip@touch-pad';
export * from '../_size/Tooltip_size_l';
export * from '../_size/Tooltip_size_m';
export * from '../_size/Tooltip_size_s';
export * from '../_view/Tooltip_view_default';
