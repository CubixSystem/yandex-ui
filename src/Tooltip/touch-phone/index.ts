export * from '../Tooltip@touch-phone';
export * from '../_size/Tooltip_size_l';
export * from '../_size/Tooltip_size_m';
export * from '../_size/Tooltip_size_s';
export * from '../_view/Tooltip_view_default';
