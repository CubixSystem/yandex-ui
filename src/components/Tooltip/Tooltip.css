.Tooltip {
    font-family: var(--control-font-family);
}

.Tooltip.Popup2::before {
    content: none;
}
