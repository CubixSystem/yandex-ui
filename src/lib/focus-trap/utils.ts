import { getTabbables } from './tabbable';

const FOCUS_TRAP_GUARD_ATTR = 'focusTrapGuard';

interface ScrollableElement {
    element: HTMLElement;
    scrollTop: number;
    scrollLeft: number;
}

let supportsPreventScrollCached: boolean | null = null;
function supportsPreventScroll() {
    if (supportsPreventScrollCached === null) {
        supportsPreventScrollCached = false;

        try {
            const element = document.createElement('div');

            element.focus({
                get preventScroll() {
                    supportsPreventScrollCached = true;
                    return true;
                },
            });
        } catch (e) {
            // ignore
        }
    }

    return supportsPreventScrollCached;
}

function getScrollableElements(element: HTMLElement): ScrollableElement[] {
    const scrollableElements: ScrollableElement[] = [];
    const rootScrollingElement = document.scrollingElement || document.documentElement;

    let parent = element.parentNode;
    while (parent instanceof HTMLElement && parent !== rootScrollingElement) {
        if (parent.offsetHeight < parent.scrollHeight || parent.offsetWidth < parent.scrollWidth) {
            scrollableElements.push({
                element: parent,
                scrollTop: parent.scrollTop,
                scrollLeft: parent.scrollLeft,
            });
        }
        parent = parent.parentNode;
    }

    if (rootScrollingElement instanceof HTMLElement) {
        scrollableElements.push({
            element: rootScrollingElement,
            scrollTop: rootScrollingElement.scrollTop,
            scrollLeft: rootScrollingElement.scrollLeft,
        });
    }

    return scrollableElements;
}

function restoreScrollPosition(scrollableElements: ScrollableElement[]) {
    for (const { element, scrollTop, scrollLeft } of scrollableElements) {
        element.scrollTop = scrollTop;
        element.scrollLeft = scrollLeft;
    }
}

export function getActiveElement() {
    return document.activeElement as HTMLElement | null;
}

export function isElementInScope(scope: Element, element: Element) {
    return scope.contains(element);
}

export function isFocusFree() {
    return getActiveElement() === document.body;
}

export function createGuard(tabIndex: number) {
    const element = document.createElement('span');
    element.tabIndex = tabIndex;
    Object.assign(element.style, {
        width: '1px',
        height: 0,
        padding: 0,
        overflow: 'hidden',
        position: 'fixed',
        top: '1px',
        left: '1px',
    });
    element.setAttribute('aria-hidden', 'true');
    element.dataset[FOCUS_TRAP_GUARD_ATTR] = 'true';

    return element;
}

export function isGuard(node: HTMLElement) {
    return Boolean(node.dataset[FOCUS_TRAP_GUARD_ATTR]);
}

export function removeElementFromDocument(element: Element) {
    if (!element.parentNode) {
        return;
    }

    element.parentNode.removeChild(element);
}

function focusWithoutScrolling(element: HTMLElement) {
    if (supportsPreventScroll()) {
        element.focus({ preventScroll: true });
    } else {
        const scrollableElements = getScrollableElements(element);
        element.focus();
        restoreScrollPosition(scrollableElements);
    }
}

export function focusElement(element?: HTMLElement | null, preventScroll?: boolean) {
    if (!element) {
        return;
    }

    try {
        if (preventScroll) {
            focusWithoutScrolling(element);
        } else {
            element.focus();
        }
    } catch (e) {
        // ignore
    }
}

export function focusFirstIn(scope: HTMLElement) {
    const [first] = getTabbables(scope, (node) => !isGuard(node));

    focusElement(first);
}

export function getNextTabbableIndex(current: number, count: number, step: number) {
    if (current === -1) {
        return 0;
    }

    return count > 0 ? (count + current + step) % count : -1;
}
