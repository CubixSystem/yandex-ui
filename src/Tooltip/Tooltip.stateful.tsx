import React, { FC, cloneElement, ReactElement, useRef, ReactNode, Children } from 'react';

import { mergeRefs } from '../lib/mergeRefs';
import { forkFn } from '../lib/forkFn';
import { useUniqId } from '../useUniqId';
import { TooltipProps as TooltipStatelessProps, Tooltip } from './Tooltip';
import { useTooltipState } from './hooks/useTooltipState';

type TooltipProps = Omit<TooltipStatelessProps, 'visible' | 'anchor'>;
type Trigger = 'click' | 'hover' | 'focus';

export type TooltipStatefulProps = TooltipProps & {
    /**
     * Компонент или элемент для которого нужно показать подсказку
     */
    children: ReactElement;

    /**
     * Содержимое, которое будет показано в подсказке
     */
    content: ReactNode;

    /**
     * Режим для показа подсказки. Может содержать несколько значений
     *
     * @default 'hover'
     */
    trigger?: Trigger | Trigger[];

    /**
     * Показывает подсказку сразу видимой
     *
     * @default false
     */
    defaultVisible?: boolean;

    /**
     * Задержка в `ms` перед открытием подсказки
     *
     * @default 0
     */
    openDelay?: number;

    /**
     * Задержка в `ms` перед закрытием подсказки
     *
     * @default 0
     */
    closeDelay?: number;
};

/**
 * Компонент используется для создания всплывающих подсказок.
 *
 * @param {TooltipStatefulProps} props Свойства компонента.
 */
export const TooltipStateful: FC<TooltipStatefulProps> = ({
    children,
    content,
    defaultVisible = false,
    id: propId,
    trigger = 'hover',
    openDelay,
    closeDelay,
    ...props
}) => {
    const uniqId = useUniqId('tooltip');
    const id = propId || uniqId;
    const { visible, open, close, toggle } = useTooltipState({ openDelay, closeDelay, defaultVisible });
    const anchorRef = useRef<HTMLElement>(null);

    const child = Children.only(children);

    // В дальнейшем данный код можно вынести куда-то выше и использовать по аналогии с Popup и Modal.
    const triggers = Array.isArray(trigger) ? trigger : [trigger];
    const handlers = triggers.reduce<any>((acc, trigger) => {
        switch (trigger) {
            case 'click':
                acc.onClick = forkFn(child.props.onClick, () => toggle());
                break;
            case 'focus':
                acc.onFocus = forkFn(child.props.onFocus, () => open());
                acc.onBlur = forkFn(child.props.onBlur, () => close());
                break;
            case 'hover':
                acc.onMouseEnter = forkFn(child.props.onMouseEnter, () => open());
                acc.onMouseLeave = forkFn(child.props.onMouseLeave, () => close());
                break;
        }
        return acc;
    }, {});

    // Определяем тип ссылки в зависимости от типа потомка,
    // нативные элементы реализуют ref, лего-компоненты должны реализовывать innerRef.
    const refKey = typeof child.type === 'string' ? 'ref' : 'innerRef';
    const ref = child.props[refKey] === undefined ? anchorRef : mergeRefs(child.props[refKey], anchorRef);

    return (
        <>
            {cloneElement(child, {
                ...handlers,
                [refKey]: ref,
                ['aria-describedby']: id,
            })}
            <Tooltip {...props} id={id} visible={visible} anchor={anchorRef} onClose={close}>
                {content}
            </Tooltip>
        </>
    );
};
