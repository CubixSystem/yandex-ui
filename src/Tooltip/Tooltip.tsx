import React, { FC } from 'react';
import { cn } from '@bem-react/classname';
import { useComponentRegistry } from '@bem-react/di';

import { IPopupProps, Direction } from '../Popup/Popup';
import { IPopupTargetAnchorProps } from '../Popup/_target/Popup_target_anchor';
import { TooltipBackdrop as Backdrop } from './Backdrop/Tooltip-Backdrop';
import { TooltipRegistry } from './Tooltip.registry/interface';
import './Tooltip.css';

export { Direction } from '../Popup/Popup';

export const cnTooltip = cn('Tooltip');

type PartialPopupProps = Pick<
    IPopupProps,
    'keepMounted' | 'className' | 'hasTail' | 'innerRef' | 'zIndex' | 'visible' | 'scope' | 'onClose' | 'onClick'
>;

type PartialPopupTargetAnchorProps = Pick<
    IPopupTargetAnchorProps,
    'mainOffset' | 'secondaryOffset' | 'tailOffset' | 'anchor'
>;

export type TooltipProps = PartialPopupProps &
    PartialPopupTargetAnchorProps & {
        /**
         * Визуальное состояние подсказки
         */
        state?: 'warning' | 'alert' | 'success';

        /**
         * Направление для раскрытия подсказки
         */
        direction?: Direction | Direction[];

        /**
         * Уникальный id подсказки
         */
        id?: string;
    };

/**
 * Компонент используется для создания всплывающих подсказок.
 *
 * @param {TooltipProps} props Свойства компонента.
 */
export const Tooltip: FC<TooltipProps> = ({
    anchor,
    children,
    className,
    direction,
    hasTail,
    id,
    innerRef,
    keepMounted,
    mainOffset,
    scope,
    secondaryOffset,
    state,
    tailOffset,
    visible,
    zIndex,
    onClose,
    onClick,
}) => {
    const defaultMainOffset = hasTail ? 0 : 4;
    const { Popup } = useComponentRegistry<TooltipRegistry>(cnTooltip());
    const directions = Array.isArray(direction) ? direction : [direction as Direction];

    return (
        <Popup
            onClose={onClose}
            onClick={onClick}
            anchor={anchor}
            className={cnTooltip({ visible, state }, [className])}
            directions={direction ? directions : undefined}
            hasTail={hasTail}
            innerRef={innerRef}
            keepMounted={keepMounted}
            mainOffset={mainOffset || defaultMainOffset}
            scope={scope}
            secondaryOffset={secondaryOffset}
            tailOffset={tailOffset}
            target="anchor"
            view="default"
            visible={visible}
            zIndex={zIndex}
            // prettier-ignore
            unstable_onRenderTail={(tail) => (
                <Backdrop>{hasTail && tail}</Backdrop>
            )}
        >
            <div role="tooltip" id={id} className={cnTooltip('Content')}>
                {children}
            </div>
        </Popup>
    );
};

Tooltip.displayName = 'Tooltip';
