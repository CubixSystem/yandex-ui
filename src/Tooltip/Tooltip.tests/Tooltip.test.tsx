import React, { ComponentType, createRef } from 'react';
import { ReactWrapper, mount, render } from 'enzyme';
import { ExtractProps } from '@bem-react/core';

import { Tooltip as TooltipCommon } from '../Tooltip';
import { Tooltip as TooltipDesktop } from '../Tooltip@desktop';
import { Tooltip as TooltipTouchPad } from '../Tooltip@touch-phone';
import { Tooltip as TooltipTouchPhone } from '../Tooltip@touch-pad';
import { serverEnvironmentSetup, delay } from '../../internal/utils';
import { Nullable } from '../../typings/utility-types';

// prettier-ignore
const platforms = [
    ['desktop', TooltipDesktop],
    ['touch-pad', TooltipTouchPad],
    ['touch-phone', TooltipTouchPhone],
];

type TooltipProps = ExtractProps<typeof TooltipDesktop>;

describe.each<any>(platforms)('Tooltip@%s', (_platform, Tooltip: ComponentType<TooltipProps>) => {
    let wrapper: Nullable<ReactWrapper> = null;

    afterEach(() => {
        if (wrapper !== null && wrapper.length > 0) {
            wrapper.unmount();
        }
    });

    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Tooltip className="mix" anchor={createRef()}>content</Tooltip>)).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            wrapper = mount(<Tooltip visible anchor={createRef()}>Добавить в избранное</Tooltip>);
            expect(wrapper).toMatchSnapshot();
        });

        test('должен быть установлен корректный displayName', () => {
            expect(TooltipCommon.displayName).toBe('Tooltip');
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLDivElement>();
            wrapper = mount(<Tooltip visible innerRef={innerRef} anchor={createRef()} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('должен устанавливать модификатор visible если открыт при инициализации', () => {
            wrapper = mount(<Tooltip visible anchor={createRef()} />);
            expect(
                wrapper
                    .find('.Tooltip')
                    .at(0)
                    .hasClass('Tooltip_visible'),
            ).toEqual(true);
        });

        test('должен устанавливать/удалять модификатор visible при открытии/закрытии', () => {
            wrapper = mount(<Tooltip anchor={createRef()} />);
            wrapper.setProps({ visible: true });
            expect(
                wrapper
                    .find('.Tooltip')
                    .at(0)
                    .hasClass('Tooltip_visible'),
            ).toEqual(true);
            wrapper.setProps({ visible: false });
            expect(
                wrapper
                    .find('.Tooltip')
                    .at(0)
                    .hasClass('Tooltip_visible'),
            ).toEqual(false);
        });
    });
});
