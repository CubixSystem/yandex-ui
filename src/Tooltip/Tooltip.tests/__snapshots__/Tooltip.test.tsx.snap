// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<RegistryResolver(Tooltip)
  anchor={
    Object {
      "current": null,
    }
  }
  visible={true}
>
  <Tooltip
    anchor={
      Object {
        "current": null,
      }
    }
    visible={true}
  >
    <SimpleComposeWrapper [view:default]
      anchor={
        Object {
          "current": null,
        }
      }
      className="Tooltip Tooltip_visible"
      mainOffset={4}
      target="anchor"
      unstable_onRenderTail={[Function]}
      view="default"
      visible={true}
    >
      <WithBemMod(Popup2)[target:anchor]
        anchor={
          Object {
            "current": null,
          }
        }
        className="Popup2 Popup2_view_default Tooltip Tooltip_visible"
        mainOffset={4}
        target="anchor"
        unstable_onRenderTail={[Function]}
        view="default"
        visible={true}
      >
        <WithBemModEnhance(RegistryResolver(Popup2))
          anchor={
            Object {
              "current": null,
            }
          }
          className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
          directions={
            Array [
              "bottom-left",
              "bottom-center",
              "bottom-right",
              "top-left",
              "top-center",
              "top-right",
              "right-top",
              "right-center",
              "right-bottom",
              "left-top",
              "left-center",
              "left-bottom",
            ]
          }
          mainOffset={4}
          scope={
            Object {
              "current": <body>
                <div
                  class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                  style="top: -9999px; left: -9999px;"
                >
                  <div
                    class="Tooltip-Content"
                    role="tooltip"
                  >
                    Добавить в избранное
                  </div>
                  <div
                    class="Tooltip-Backdrop"
                  />
                </div>
              </body>,
            }
          }
          secondaryOffset={0}
          tailOffset={0}
          target="anchor"
          unstable_onRenderTail={[Function]}
          view="default"
          viewportOffset={0}
          visible={true}
        >
          <RegistryResolver(Popup2)
            className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
            direction="bottom-left"
            innerRef={[Function]}
            position={
              Object {
                "left": -9999,
                "top": -9999,
              }
            }
            scope={
              Object {
                "current": <body>
                  <div
                    class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                    style="top: -9999px; left: -9999px;"
                  >
                    <div
                      class="Tooltip-Content"
                      role="tooltip"
                    >
                      Добавить в избранное
                    </div>
                    <div
                      class="Tooltip-Backdrop"
                    />
                  </div>
                </body>,
              }
            }
            tailRef={[Function]}
            unstable_essentialRefs={
              Array [
                Object {
                  "current": null,
                },
              ]
            }
            unstable_onRenderTail={[Function]}
            view="default"
            visible={true}
          >
            <Popup2
              className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
              direction="bottom-left"
              innerRef={[Function]}
              position={
                Object {
                  "left": -9999,
                  "top": -9999,
                }
              }
              scope={
                Object {
                  "current": <body>
                    <div
                      class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                      style="top: -9999px; left: -9999px;"
                    >
                      <div
                        class="Tooltip-Content"
                        role="tooltip"
                      >
                        Добавить в избранное
                      </div>
                      <div
                        class="Tooltip-Backdrop"
                      />
                    </div>
                  </body>,
                }
              }
              tailRef={[Function]}
              unstable_essentialRefs={
                Array [
                  Object {
                    "current": null,
                  },
                ]
              }
              unstable_onRenderTail={[Function]}
              view="default"
              visible={true}
            >
              <Portal
                containerInfo={
                  <body>
                    <div
                      class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                      style="top: -9999px; left: -9999px;"
                    >
                      <div
                        class="Tooltip-Content"
                        role="tooltip"
                      >
                        Добавить в избранное
                      </div>
                      <div
                        class="Tooltip-Backdrop"
                      />
                    </div>
                  </body>
                }
              >
                <LayerManager
                  essentialRefs={
                    Array [
                      Object {
                        "current": <div
                          class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                          style="top: -9999px; left: -9999px;"
                        >
                          <div
                            class="Tooltip-Content"
                            role="tooltip"
                          >
                            Добавить в избранное
                          </div>
                          <div
                            class="Tooltip-Backdrop"
                          />
                        </div>,
                      },
                      Object {
                        "current": null,
                      },
                    ]
                  }
                  onClose={[Function]}
                  visible={true}
                >
                  <div
                    className="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                    style={
                      Object {
                        "left": -9999,
                        "top": -9999,
                        "zIndex": undefined,
                      }
                    }
                  >
                    <div
                      className="Tooltip-Content"
                      role="tooltip"
                    >
                      Добавить в избранное
                    </div>
                    <Component>
                      <div
                        className="Tooltip-Backdrop"
                      />
                    </Component>
                  </div>
                </LayerManager>
              </Portal>
            </Popup2>
          </RegistryResolver(Popup2)>
        </WithBemModEnhance(RegistryResolver(Popup2))>
      </WithBemMod(Popup2)[target:anchor]>
    </SimpleComposeWrapper [view:default]>
  </Tooltip>
</RegistryResolver(Tooltip)>
`;

exports[`Tooltip@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `null`;

exports[`Tooltip@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<RegistryResolver(Tooltip)
  anchor={
    Object {
      "current": null,
    }
  }
  visible={true}
>
  <Tooltip
    anchor={
      Object {
        "current": null,
      }
    }
    visible={true}
  >
    <SimpleComposeWrapper [view:default]
      anchor={
        Object {
          "current": null,
        }
      }
      className="Tooltip Tooltip_visible"
      mainOffset={4}
      target="anchor"
      unstable_onRenderTail={[Function]}
      view="default"
      visible={true}
    >
      <WithBemMod(Popup2)[target:anchor]
        anchor={
          Object {
            "current": null,
          }
        }
        className="Popup2 Popup2_view_default Tooltip Tooltip_visible"
        mainOffset={4}
        target="anchor"
        unstable_onRenderTail={[Function]}
        view="default"
        visible={true}
      >
        <WithBemModEnhance(RegistryResolver(Popup2))
          anchor={
            Object {
              "current": null,
            }
          }
          className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
          directions={
            Array [
              "bottom-left",
              "bottom-center",
              "bottom-right",
              "top-left",
              "top-center",
              "top-right",
              "right-top",
              "right-center",
              "right-bottom",
              "left-top",
              "left-center",
              "left-bottom",
            ]
          }
          mainOffset={4}
          scope={
            Object {
              "current": <body>
                <div
                  class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                  style="top: -9999px; left: -9999px;"
                >
                  <div
                    class="Tooltip-Content"
                    role="tooltip"
                  >
                    Добавить в избранное
                  </div>
                  <div
                    class="Tooltip-Backdrop"
                  />
                </div>
              </body>,
            }
          }
          secondaryOffset={0}
          tailOffset={0}
          target="anchor"
          unstable_onRenderTail={[Function]}
          view="default"
          viewportOffset={0}
          visible={true}
        >
          <RegistryResolver(Popup2)
            className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
            direction="bottom-left"
            innerRef={[Function]}
            position={
              Object {
                "left": -9999,
                "top": -9999,
              }
            }
            scope={
              Object {
                "current": <body>
                  <div
                    class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                    style="top: -9999px; left: -9999px;"
                  >
                    <div
                      class="Tooltip-Content"
                      role="tooltip"
                    >
                      Добавить в избранное
                    </div>
                    <div
                      class="Tooltip-Backdrop"
                    />
                  </div>
                </body>,
              }
            }
            tailRef={[Function]}
            unstable_essentialRefs={
              Array [
                Object {
                  "current": null,
                },
              ]
            }
            unstable_onRenderTail={[Function]}
            view="default"
            visible={true}
          >
            <Popup2
              className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
              direction="bottom-left"
              innerRef={[Function]}
              position={
                Object {
                  "left": -9999,
                  "top": -9999,
                }
              }
              scope={
                Object {
                  "current": <body>
                    <div
                      class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                      style="top: -9999px; left: -9999px;"
                    >
                      <div
                        class="Tooltip-Content"
                        role="tooltip"
                      >
                        Добавить в избранное
                      </div>
                      <div
                        class="Tooltip-Backdrop"
                      />
                    </div>
                  </body>,
                }
              }
              tailRef={[Function]}
              unstable_essentialRefs={
                Array [
                  Object {
                    "current": null,
                  },
                ]
              }
              unstable_onRenderTail={[Function]}
              view="default"
              visible={true}
            >
              <Portal
                containerInfo={
                  <body>
                    <div
                      class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                      style="top: -9999px; left: -9999px;"
                    >
                      <div
                        class="Tooltip-Content"
                        role="tooltip"
                      >
                        Добавить в избранное
                      </div>
                      <div
                        class="Tooltip-Backdrop"
                      />
                    </div>
                  </body>
                }
              >
                <LayerManager
                  essentialRefs={
                    Array [
                      Object {
                        "current": <div
                          class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                          style="top: -9999px; left: -9999px;"
                        >
                          <div
                            class="Tooltip-Content"
                            role="tooltip"
                          >
                            Добавить в избранное
                          </div>
                          <div
                            class="Tooltip-Backdrop"
                          />
                        </div>,
                      },
                      Object {
                        "current": null,
                      },
                    ]
                  }
                  onClose={[Function]}
                  visible={true}
                >
                  <div
                    className="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                    style={
                      Object {
                        "left": -9999,
                        "top": -9999,
                        "zIndex": undefined,
                      }
                    }
                  >
                    <div
                      className="Tooltip-Content"
                      role="tooltip"
                    >
                      Добавить в избранное
                    </div>
                    <Component>
                      <div
                        className="Tooltip-Backdrop"
                      />
                    </Component>
                  </div>
                </LayerManager>
              </Portal>
            </Popup2>
          </RegistryResolver(Popup2)>
        </WithBemModEnhance(RegistryResolver(Popup2))>
      </WithBemMod(Popup2)[target:anchor]>
    </SimpleComposeWrapper [view:default]>
  </Tooltip>
</RegistryResolver(Tooltip)>
`;

exports[`Tooltip@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `null`;

exports[`Tooltip@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<RegistryResolver(Tooltip)
  anchor={
    Object {
      "current": null,
    }
  }
  visible={true}
>
  <Tooltip
    anchor={
      Object {
        "current": null,
      }
    }
    visible={true}
  >
    <SimpleComposeWrapper [view:default]
      anchor={
        Object {
          "current": null,
        }
      }
      className="Tooltip Tooltip_visible"
      mainOffset={4}
      target="anchor"
      unstable_onRenderTail={[Function]}
      view="default"
      visible={true}
    >
      <WithBemMod(Popup2)[target:anchor]
        anchor={
          Object {
            "current": null,
          }
        }
        className="Popup2 Popup2_view_default Tooltip Tooltip_visible"
        mainOffset={4}
        target="anchor"
        unstable_onRenderTail={[Function]}
        view="default"
        visible={true}
      >
        <WithBemModEnhance(RegistryResolver(Popup2))
          anchor={
            Object {
              "current": null,
            }
          }
          className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
          directions={
            Array [
              "bottom-left",
              "bottom-center",
              "bottom-right",
              "top-left",
              "top-center",
              "top-right",
              "right-top",
              "right-center",
              "right-bottom",
              "left-top",
              "left-center",
              "left-bottom",
            ]
          }
          mainOffset={4}
          scope={
            Object {
              "current": <body>
                <div
                  class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                  style="top: -9999px; left: -9999px;"
                >
                  <div
                    class="Tooltip-Content"
                    role="tooltip"
                  >
                    Добавить в избранное
                  </div>
                  <div
                    class="Tooltip-Backdrop"
                  />
                </div>
              </body>,
            }
          }
          secondaryOffset={0}
          tailOffset={0}
          target="anchor"
          unstable_onRenderTail={[Function]}
          view="default"
          viewportOffset={0}
          visible={true}
        >
          <RegistryResolver(Popup2)
            className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
            direction="bottom-left"
            innerRef={[Function]}
            position={
              Object {
                "left": -9999,
                "top": -9999,
              }
            }
            scope={
              Object {
                "current": <body>
                  <div
                    class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                    style="top: -9999px; left: -9999px;"
                  >
                    <div
                      class="Tooltip-Content"
                      role="tooltip"
                    >
                      Добавить в избранное
                    </div>
                    <div
                      class="Tooltip-Backdrop"
                    />
                  </div>
                </body>,
              }
            }
            tailRef={[Function]}
            unstable_essentialRefs={
              Array [
                Object {
                  "current": null,
                },
              ]
            }
            unstable_onRenderTail={[Function]}
            view="default"
            visible={true}
          >
            <Popup2
              className="Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
              direction="bottom-left"
              innerRef={[Function]}
              position={
                Object {
                  "left": -9999,
                  "top": -9999,
                }
              }
              scope={
                Object {
                  "current": <body>
                    <div
                      class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                      style="top: -9999px; left: -9999px;"
                    >
                      <div
                        class="Tooltip-Content"
                        role="tooltip"
                      >
                        Добавить в избранное
                      </div>
                      <div
                        class="Tooltip-Backdrop"
                      />
                    </div>
                  </body>,
                }
              }
              tailRef={[Function]}
              unstable_essentialRefs={
                Array [
                  Object {
                    "current": null,
                  },
                ]
              }
              unstable_onRenderTail={[Function]}
              view="default"
              visible={true}
            >
              <Portal
                containerInfo={
                  <body>
                    <div
                      class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                      style="top: -9999px; left: -9999px;"
                    >
                      <div
                        class="Tooltip-Content"
                        role="tooltip"
                      >
                        Добавить в избранное
                      </div>
                      <div
                        class="Tooltip-Backdrop"
                      />
                    </div>
                  </body>
                }
              >
                <LayerManager
                  essentialRefs={
                    Array [
                      Object {
                        "current": <div
                          class="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                          style="top: -9999px; left: -9999px;"
                        >
                          <div
                            class="Tooltip-Content"
                            role="tooltip"
                          >
                            Добавить в избранное
                          </div>
                          <div
                            class="Tooltip-Backdrop"
                          />
                        </div>,
                      },
                      Object {
                        "current": null,
                      },
                    ]
                  }
                  onClose={[Function]}
                  visible={true}
                >
                  <div
                    className="Popup2 Popup2_visible Popup2_direction_bottom-left Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
                    style={
                      Object {
                        "left": -9999,
                        "top": -9999,
                        "zIndex": undefined,
                      }
                    }
                  >
                    <div
                      className="Tooltip-Content"
                      role="tooltip"
                    >
                      Добавить в избранное
                    </div>
                    <Component>
                      <div
                        className="Tooltip-Backdrop"
                      />
                    </Component>
                  </div>
                </LayerManager>
              </Portal>
            </Popup2>
          </RegistryResolver(Popup2)>
        </WithBemModEnhance(RegistryResolver(Popup2))>
      </WithBemMod(Popup2)[target:anchor]>
    </SimpleComposeWrapper [view:default]>
  </Tooltip>
</RegistryResolver(Tooltip)>
`;

exports[`Tooltip@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `null`;
